lib/Storage.js

```diff
@@ -1,87 +1,132 @@
+XPCOMUtils.defineLazyModuleGetter(this, 'OS',
+                                  'resource://gre/modules/osfile.jsm'); /* global OS */
+XPCOMUtils.defineLazyModuleGetter(this, 'Task',
+                                  'resource://gre/modules/Task.jsm'); /* global Task */
 
-    request.onupgradeneeded = event => {
-      var db = event.target.result;
-      db.createObjectStore('storage', {keyPath: 'key'});
-    };
-  });
+const {Log} = require('./Log.js');
+
+let storePromise;
+
+function loadStorage() {
+  if (storePromise === undefined) {
+    let path = OS.Path.join(OS.Constants.Path.profileDir, 'recipe-client.json');
+    let storage = new JSONFile({path});
+    storePromise = Task.spawn(function* () {
+      yield storage.load();
+      return storage;
+    });
+  }
+  return storePromise;
+}
 
-  function makeKey(suffix) {
-    return `${prefix}.${suffix}`;
+exports.Storage = function(prefix, sandbox) {
+  if (!sandbox) {
+    throw new Error('No sandbox passed');
   }
 
   const storageInterface = {
+    /**
+     * Sets an item in the prefixed storage.
+     * @returns {Promise}
+     * @resolves With the stored value, or null.
+     * @rejects Javascript exception.
+     */
     getItem(keySuffix) {
-      return idbPromise.then(db => {
-        return new sandbox.Promise((resolve, reject) => {
-          let key = makeKey(keySuffix);
-          let transaction = db.transaction(['storage'], 'readonly');
-          let request = transaction.objectStore('storage').get(key);
-          request.onsuccess = event => {
-            if (event.target.result) {
-              resolve(event.target.result.value);
-            } else {
-              resolve(null);
-            }
-          };
-          request.onerror = err => reject(err);
+      return loadStorage()
+        .then(store => {
+          let namespace = store.data[prefix] || {};
+          const value = namespace[keySuffix] || null;
+          return sandbox.Promise.resolve(Cu.cloneInto(value, sandbox));
+        })
+        .catch(err => {
+          Log.error(err);
+          throw new Error(err.toString());
         });
-      });
     },
 
+    /**
+     * Sets an item in the prefixed storage.
+     * @returns {Promise}
+     * @resolves When the operation is completed succesfully
+     * @rejects Javascript exception.
+     */
     setItem(keySuffix, value) {
-      return idbPromise.then(db => {
-        return new sandbox.Promise((resolve, reject) => {
-          let key = makeKey(keySuffix);
-          let transaction = db.transaction(['storage'], 'readwrite');
-          let request = transaction.objectStore('storage').put({key, value});
-          request.onsuccess = () => resolve();
-          request.onerror = err => reject(err);
+      return loadStorage()
+        .then(store => {
+          if (!(prefix in store.data)) {
+            store.data[prefix] = {};
+          }
+          store.data[prefix][keySuffix] = value;
+          return store.save();
```

Comment:
So [save just got renamed to _save yesterday](https://hg.mozilla.org/mozilla-central/rev/170c0f7ef959). Seems like it's not intended to be public.

Also, I'm not convinced that waiting for the data to be properly saved is needed. `JSONFile.jsm` ensures that the data is saved before Firefox shuts down, so we don't need to worry about Firefox closing before the write happens. And I don't think we're going to have a lot of smart failure handlers if writing to the file fails, or at least until we have one we should go with the recommended path. There's also the question of if this code will run on the main thread or not (did we figure out if the sandbox is off the main thread? I wonder if that applies to unsandboxed code called by the sandbox); if so, then we _really_ don't want to wait for a file write.